File: backend/src/routes/documents.ts
import express from 'express';
import multer from 'multer';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs/promises';
import { authenticateToken, type AuthRequest } from '../middleware/auth.js';
import { db } from '../database/schema.js';
import { parseDocument } from '../services/documentParser.js';
import { analyzeCompliance } from '../services/complianceAnalyzer.js';
import type { Document, Issue } from '../types/index.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const router = express.Router();

// Configure multer for file uploads
const storage = multer.diskStorage({
  destination: async (req, file, cb) => {
    const uploadDir = path.join(__dirname, '../../uploads');
    await fs.mkdir(uploadDir, { recursive: true });
    cb(null, uploadDir);
  },
  filename: (req, file, cb) => {
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1e9);
    cb(null, uniqueSuffix + '-' + file.originalname);
  },
});

const upload = multer({
  storage,
  limits: { fileSize: 10 * 1024 * 1024 }, // 10MB
  fileFilter: (req, file, cb) => {
    const allowedTypes = [
      'application/pdf',
      'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
      'application/msword',
      'text/plain',
    ];
    if (allowedTypes.includes(file.mimetype)) {
      cb(null, true);
    } else {
      cb(new Error('Invalid file type. Only PDF, Word, and text files are allowed.'));
    }
  },
});

// Upload and analyze document
router.post('/upload', authenticateToken, upload.single('file'), async (req: AuthRequest, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }

    const { title } = req.body;
    if (!title) {
      return res.status(400).json({ error: 'Document title is required' });
    }

    const userId = req.userId!;
    const filePath = req.file.path;
    const fileName = req.file.originalname;
    const fileType = req.file.mimetype;

    // Insert document record
    const insertDoc = db.prepare(
      `INSERT INTO documents (user_id, title, file_name, file_type, file_path, content, status)
       VALUES (?, ?, ?, ?, ?, ?, ?)`
    );

    const result = insertDoc.run(userId, title, fileName, fileType, filePath, '', 'pending');
    const documentId = result.lastInsertRowid;

    // Parse document in background
    setImmediate(async () => {
      try {
        // Update status to analyzing
        db.prepare('UPDATE documents SET status = ? WHERE id = ?').run('analyzing', documentId);

        // Parse document content
        const content = await parseDocument(filePath, fileType);

        // Update document with content
        db.prepare('UPDATE documents SET content = ? WHERE id = ?').run(content, documentId);

        // Analyze compliance
        const analysis = await analyzeCompliance(content, title);

        // Insert issues
        const insertIssue = db.prepare(
          `INSERT INTO issues (document_id, severity, category, title, description, location, recommendation, jurisdiction)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
        );

        for (const issue of analysis.issues) {
          insertIssue.run(
            documentId,
            issue.severity,
            issue.category,
            issue.title,
            issue.description,
            issue.location || null,
            issue.recommendation,
            issue.jurisdiction || null
          );
        }

        // Update document status
        db.prepare('UPDATE documents SET status = ?, analyzed_at = CURRENT_TIMESTAMP WHERE id = ?').run(
          'completed',
          documentId
        );

        console.log(`Document ${documentId} analyzed successfully`);
      } catch (error) {
        console.error('Error analyzing document:', error);
        db.prepare('UPDATE documents SET status = ? WHERE id = ?').run('failed', documentId);
      }
    });

    res.json({
      id: documentId,
      message: 'Document uploaded successfully. Analysis in progress.',
    });
  } catch (error) {
    console.error('Upload error:', error);
    res.status(500).json({ error: 'Failed to upload document' });
  }
});

// Get all documents for user
router.get('/', authenticateToken, (req: AuthRequest, res) => {
  const userId = req.userId!;
  const documents = db.prepare('SELECT * FROM documents WHERE user_id = ? ORDER BY uploaded_at DESC').all(userId) as Document[];
  res.json(documents);
});

// Get single document with issues
router.get('/:id', authenticateToken, (req: AuthRequest, res) => {
  const userId = req.userId!;
  const documentId = req.params.id;

  const document = db.prepare('SELECT * FROM documents WHERE id = ? AND user_id = ?').get(documentId, userId) as Document | undefined;

  if (!document) {
    return res.status(404).json({ error: 'Document not found' });
  }

  const issues = db.prepare('SELECT * FROM issues WHERE document_id = ? ORDER BY severity, created_at').all(documentId) as Issue[];

  res.json({ ...document, issues });
});

// Delete document
router.delete('/:id', authenticateToken, async (req: AuthRequest, res) => {
  const userId = req.userId!;
  const documentId = req.params.id;

  const document = db.prepare('SELECT * FROM documents WHERE id = ? AND user_id = ?').get(documentId, userId) as Document | undefined;

  if (!document) {
    return res.status(404).json({ error: 'Document not found' });
  }

  // Delete file
  if (document.file_path) {
    try {
      await fs.unlink(document.file_path);
    } catch (error) {
      console.error('Error deleting file:', error);
    }
  }

  // Delete database record (cascade will delete issues and tasks)
  db.prepare('DELETE FROM documents WHERE id = ?').run(documentId);

  res.json({ message: 'Document deleted successfully' });
});

export default router;

File: backend/app/routes/documents.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form, BackgroundTasks
from sqlalchemy.orm import Session
from datetime import datetime
import os
import shutil

from ..database import get_db, User, Document, Issue
from ..schemas import Document as DocumentSchema, DocumentWithIssues
from ..auth import get_current_user
from ..services.document_parser import parse_document
from ..services.compliance_analyzer import analyze_compliance

router = APIRouter()

UPLOAD_DIR = "uploads"
os.makedirs(UPLOAD_DIR, exist_ok=True)

async def process_document(document_id: int, file_path: str, file_type: str, title: str, db_session):
    """Background task to analyze document"""
    try:
        # Update status to analyzing
        doc = db_session.query(Document).filter(Document.id == document_id).first()
        if doc:
            doc.status = "analyzing"
            db_session.commit()

        # Parse document
        content = parse_document(file_path, file_type)

        # Update content
        doc = db_session.query(Document).filter(Document.id == document_id).first()
        if doc:
            doc.content = content
            db_session.commit()

        # Analyze compliance
        analysis = await analyze_compliance(content, title)

        # Insert issues
        for issue_data in analysis["issues"]:
            issue = Issue(
                document_id=document_id,
                severity=issue_data["severity"],
                category=issue_data["category"],
                title=issue_data["title"],
                description=issue_data["description"],
                location=issue_data.get("location"),
                recommendation=issue_data["recommendation"],
                jurisdiction=issue_data.get("jurisdiction")
            )
            db_session.add(issue)

        # Update document status
        doc = db_session.query(Document).filter(Document.id == document_id).first()
        if doc:
            doc.status = "completed"
            doc.analyzed_at = datetime.utcnow()
            db_session.commit()

        print(f"Document {document_id} analyzed successfully")

    except Exception as e:
        print(f"Error analyzing document: {e}")
        doc = db_session.query(Document).filter(Document.id == document_id).first()
        if doc:
            doc.status = "failed"
            db_session.commit()
    finally:
        db_session.close()

@router.post("/upload")
async def upload_document(
    background_tasks: BackgroundTasks,
    title: str = Form(...),
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    # Validate file type
    allowed_types = [
        "application/pdf",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "application/msword",
        "text/plain"
    ]

    if file.content_type not in allowed_types:
        raise HTTPException(
            status_code=400,
            detail="Invalid file type. Only PDF, Word, and text files are allowed."
        )

    # Save file
    file_extension = file.filename.split(".")[-1]
    unique_filename = f"{datetime.now().timestamp()}_{file.filename}"
    file_path = os.path.join(UPLOAD_DIR, unique_filename)

    with open(file_path, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)

    # Create document record
    db_document = Document(
        user_id=current_user.id,
        title=title,
        file_name=file.filename,
        file_type=file.content_type,
        file_path=file_path,
        content="",
        status="pending"
    )
    db.add(db_document)
    db.commit()
    db.refresh(db_document)

    # Start background analysis
    from ..database import SessionLocal
    background_tasks.add_task(
        process_document,
        db_document.id,
        file_path,
        file.content_type,
        title,
        SessionLocal()
    )

    return {
        "id": db_document.id,
        "message": "Document uploaded successfully. Analysis in progress."
    }

@router.get("/", response_model=list[DocumentSchema])
async def get_documents(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    documents = db.query(Document).filter(
        Document.user_id == current_user.id
    ).order_by(Document.uploaded_at.desc()).all()

    return [DocumentSchema.from_orm(doc) for doc in documents]

@router.get("/{document_id}", response_model=DocumentWithIssues)
async def get_document(
    document_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    document = db.query(Document).filter(
        Document.id == document_id,
        Document.user_id == current_user.id
    ).first()

    if not document:
        raise HTTPException(status_code=404, detail="Document not found")

    return DocumentWithIssues.from_orm(document)

@router.delete("/{document_id}")
async def delete_document(
    document_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    document = db.query(Document).filter(
        Document.id == document_id,
        Document.user_id == current_user.id
    ).first()

    if not document:
        raise HTTPException(status_code=404, detail="Document not found")

    # Delete file
    if document.file_path and os.path.exists(document.file_path):
        os.remove(document.file_path)

    # Delete database record
    db.delete(document)
    db.commit()

    return {"message": "Document deleted successfully"}
